R/conditional-transform.R

```diff
@@ -0,0 +1,101 @@
+#'
+#' `conditionalTransform` is similar to `makeCaseVariable` however `conditionalTransform` can use other crunch variables as a source of a variable, wehreas, whereas `makeCaseVariable` can only use characters. This additional power comes at a cost: `makeCaseVariable` can be executed entirely on Crunch servers, so no data needs to be downloaded or uploaded to/from the local R session. `conditionalTransform` on the other hand will download the data necesary to construct the new variable.
+#'
+#' @param ... a list of cases to evaluate
+#' @param default_value a default value to use if none of the conditions are true (default: `NA`)
+#' @param type a character that is either "categorical", "text", "numeric" what type of output should be returned?
+#' @param categories if `type="categorical"`, these are all of the categories that should be in the resulting variable
+#'
+#' @return either a factor (if `type="categorical"`) or character (if `type="text"`)
+#' @examples
+#' \dontrun{
+#' ds$cat_opinion <- conditionalTransform(ds$pet1 == 'Cat' ~ ds$Opinion1,
+#'                                        ds$pet2 == 'Cat' ~ ds$Opinion2,
+#'                                        ds$pet3 == 'Cat' ~ ds$Opinion3)
+#' ds$dot_opinion <- conditionalTransform(ds$pet1 == 'Dog' ~ ds$Opinion1,
+#'                                        ds$pet2 == 'Dog' ~ ds$Opinion2,
+#'                                        ds$pet3 == 'Dog' ~ ds$Opinion3)
+#' }
+#'
+#' @export
+conditionalTransform <- function (..., default_value=NA, type="categorical", categories=NULL) {
+    formulas <- list(...)
+    n <- length(formulas)
+
+    if (n == 0) {
+        halt("no conditions have been supplied")
+    }
+    if (!type %in% c("categorical", "text", "numeric")){
+        halt("type must be either ", dQuote("categorical"), dQuote("text"), " or ", dQuote("numeric"))
+    }
+    if (type != "categorical" & !is.null(categories)){
+        halt("type is not ", dQuote("categorical"), " ignoring ", dQuote("categories"))
+    }
+
+    cases <- vector("list", n)
+    values <- vector("list", n)
+    for (i in seq_len(n)) {
```

Comment:
Why are the existing formula evaluation functions not adequate here? Perhaps those should be refactored so that all of the formula evaluation is consolidated.

Also, looking at this reminds me: perhaps this function, and for that matter, `makeCaseVariable` too, should take a `data` argument in which the formula should be evaluated. `function(formula, data)` is the usual way of dealing with formulas. That way the `$` can be dropped from all of those case definitions.